### PR TITLE
docs: deprecate `ComponentFactoryResolver` class

### PR DESCRIPTION
**Pull request by AndrewKushnir**
NOTE: this PR is currently blocked by #43022, which would bring support for Component Types to the `ViewContainerRef.createComponent` function.

With Ivy, there is no need to resolve Component factories. Framework APIs allow to use Component Types directly. Since ViewEngine was deprecated, the `ComponentFactoryResolver` class is now deprecated too. The `NgModuleRef.componentFactoryResolver` function is also deprecated, since it's just a helper method to create an instance of a `ComponentFactoryResolver` class in a context of an NgModule.

No replacement is needed, use Framework APIs that allow passing Component Types directly.

## PR Type
What kind of change does this PR introduce?

- [x] Other: API deprecation.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by IgorMinar — APPROVED** [review]
LGTM for public-api but let's wait until #43022 before merging this! Thanks!

Reviewed-for: public-api

File: aio/content/guide/deprecations.md [hunk 1]
```diff
@@ -100,7 +102,9 @@ This section contains a complete list all of the currently-deprecated APIs, with
...
+| [`ComponentFactoryResolver`](api/core/ComponentFactoryResolver)                          | n/a                                                                                | v13                    | No replacement needed, use FW APIs that allow passing Component type instead of a factory function.                                                                                                                                                                                                                  |
+| [`NgModuleRef.componentFactoryResolver`](api/core/NgModuleRef#componentFactoryResolver)                          | n/a                                                                                | v13                    |  No replacement needed, use FW APIs that allow passing Component type instead of a factory function.                                                                                                                                                                                                                  |
```
Comment (jelbourn) [inline aio/content/guide/deprecations.md:107]: nit: "FW" -> "framework", and probably "Component type" -> "Component class"

File: packages/core/src/linker/component_factory_resolver.ts [hunk 2]
```diff
@@ -44,6 +44,9 @@ class _NullComponentFactoryResolver implements ComponentFactoryResolver {
  * @see <live-example name="dynamic-component-loader" noDownload></live-example>
 of the code in this cookbook
  * @publicApi
+ *
+ * @deprecated With Ivy, there is no need to resolve Component factories. No replacement is needed,
+it's possible to use Component Types directly.
```
Comment (jelbourn) [inline packages/core/src/linker/component_factory_resolver.ts:49]:
```suggestion
 * @deprecated Angular no longer requires component factories to dynamically
 * create components. All dynamic component creation APIs now directly accept a
 * component class.
```
nit: more active voice (here and elsewhere)

**Review by jelbourn — APPROVED** [review]
LGTM aside from some docs nits

File: packages/core/src/linker/component_factory_resolver.ts (hunk 2 above)
Comment (AndrewKushnir) [reply packages/core/src/linker/component_factory_resolver.ts:49]: Thanks for the feedback Jeremy! I've applied the necessary changes.

**Review by jelbourn — APPROVED** [review]
lgtm

**Review by jessicajaniuk — APPROVED** [review]
LGTM :cookie:
reviewed-for: public-api

**Review by alxhub — APPROVED** [review]
Reviewed-for: public-api

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:399337851:BASE:399337860:1632796632349:29e1b07).

**Comment by AndrewKushnir** [comment]
Further cleanup work revealed some use-cases for which we don't have a good replacement at this moment. For example:

https://github.com/angular/angular/blob/ab3de40ba3e45a0e6df30f62d41cf65ed0b021e2/aio/content/examples/elements/src/app/popup.service.ts#L19-L24

If we deprecate the `ComponentFactoryResolver` now, we'd loose an API to go from a Component class -> `ComponentRef`, which is needed in some scenarios.

Instead of the `ComponentFactoryResolver` deprecation in v13, the proposal is to revisit use-cases for `ComponentFactory` + `ComponentFactoryResolver`, come up with replacement APIs and deprecate both at the same time (in one of the next major versions after v13). Meanwhile, I'd update the `ComponentFactoryResolver` docs (and relevant guides) to mention that it's no longer needed for dynamic components use-case. This work will happen in a separate PR, so I'm closing this one for now (will reopen later).

**Pull request closed without merging**